python/singa/autograd.py

```diff
@@ -749,3 +757,89 @@ def __call__(self, x):
 
         y = conv2d(x, self.W, self.b, self.handle)
         return y
+
+
+class BatchNorm2d(NewLayer):
+
+    def __init__(self, num_features, momentum=0.9):
+        self.channels = num_features
+        self.momentum = momentum
+
+        param_shape = (self.channels,)
+
+        self.scale = Tensor(shape=param_shape,
+                            requires_grad=True, stores_grad=True)
+        self.scale.set_value(1.0)
+
+        self.bias = Tensor(shape=param_shape,
+                           requires_grad=True, stores_grad=True)
+        self.bias.set_value(0.0)
+
+        self.runningmean = Tensor(
+            shape=param_shape, requires_grad=False, stores_grad=False)
+        self.runningvariance = Tensor(
+            shape=param_shape, requires_grad=False, stores_grad=False)
+
+    def __call__(self, x):
+        assert x.shape[1] == self.channels, 'number of channels dismatched.'
+
+        self.device_check(x, self.scale, self.bias,
+                          self.runningmean, self.runningvariance)
+
+        if x.device.id() == -1:
+            raise NotImplementedError
+
+        else:
+            if not hasattr(self, 'handle'):
+                self.handle = singa.CudnnBatchNormHandle(
+                    self.momentum, x.data, self.runningmean.data, self.runningvariance.data)
+            elif x.shape[0] != self.handle.batchsize:
+                self.handle = singa.CudnnBatchNormHandle(
+                    self.momentum, x.data, self.runningmean.data, self.runningvariance.data)
+        self.handle.device_id = x.device.id()
+
+        y = batchnorm2d(x, self.scale, self.bias, self.handle)
+        return y
+
+
+class _BatchNorm2d(Operation):
+
+    def __init__(self, handle):
+        self.handle = handle
+
+    def forward(self, x, scale, bias):
+        if training:
+            resultmean = CTensor([scale.shape(0)])
```

Comment:
if you create a new resultemean and resultvariance, how do you accumulate the mean and var?

Reply:
in my former design, i let cache=(x,). then want to create resultmean and resultvariance in Forward() and push_back them to cache during the forward process. then the parameter declaration for cache cannot be set as 'const'.  then when i test, i sent GpuBatchNormForwardTraining(x, scale, bias, [x], handle), there is a argument type error. [x] cannot be understanded as vector.